### PR TITLE
Progress Bar Rewrite

### PR DESCRIPTION
**Pull request by dearmochi**
## What Does This PR Do

- **Completely rewrites progress bar code.**
  - The bar now transitions smoothly from 0% to 100% rather than in hard-drawn increments of 5%.
  - Instead, a progress bar's visual is comprised of a backdrop, fill and mask. The fill gradually grows in size as the mask shrinks using an alpha mask filter.
  - The result of this is it makes making new progress bar themes very easy, which could be used to denote actions of a certain type (cult, dangerous actions, etc.)
- Removes `progessbar.dmi` (notice the typo), replacing it with `progressbar.dmi`

## Why It's Good For The Game

Out with the old (and especially the hard-drawn progress bar, that was awful)

## Images of changes

MkGFgetnjy
_Preview_

z4lxs1iSHD
_Progress bar stacking_

image
_A progress bar theme is composed of 3 images only, making customization easy_

Bonus: [warning theme (WEBM)](https://cdn.discordapp.com/attachments/720022922177151107/819183154216697866/gRwnpzbrEu.webm)

## Changelog
:cl:
add: Add the possibility to create progress bar themes
tweak: Progress bars now transition smoothly from 0% to 100%, rather than in increments of 5%
/:cl:

### CODE REVIEW

File: code/datums/progressbar.dm [hunk 1]
```diff
@@ -1,70 +1,182 @@
...
+
+/**
+  * Updates the fill. Do your visual changes here.
+  *
+  * Arguments:
+  * * fraction - The new (value / goal) fraction.
+  */
+/datum/progress_bar/proc/update_fill(fraction)
+	var/F = fill.filters[1]
+	F?:x = -fill_width * (1 - fraction) // I know what you're thinking. I should be shot, but dreamchecker doesn't know this type for some reason yet
```
Comment (dearmochi) [inline code/datums/progressbar.dm:136]: Yes this is a colon operator, fight me

File: code/datums/progressbar.dm [hunk 2]
```diff
@@ -1,70 +1,180 @@
...
+  * Shifts the progress bar down.
+  */
+/datum/progress_bar/proc/shift_down()
+	--list_index
+	var/shift_height = holder.pixel_y - height
+	animate(holder, pixel_y = shift_height, time = 0.5 SECONDS, easing = SINE_EASING | EASE_OUT)
+
+/datum/progress_bar/default/update_fill(fraction)
 	. = ..()
+	fill.color = gradient(0, COLOR_RED, 0.25, COLOR_ORANGE, 0.5, COLOR_GREEN, 0.75, COLOR_SKY_BLUE, fraction)
```
Comment (AffectedArc07) [inline code/datums/progressbar.dm:148]: `gradient()` is a 514 proc.

File: code/datums/progressbar.dm [hunk 3]
```diff
@@ -1,70 +1,180 @@
...
+			var/datum/progress_bar/P = I
+			if(P.list_index > list_index)
+				P.shift_down()
+
+		bars -= src
+		if(!length(bars))
+			LAZYREMOVE(user.progressbars, target)
+
+	animate(holder, alpha = 0, time = 0.5 SECONDS)
+	spawn(5)
```
Comment (SteelSlayer) [inline code/datums/progressbar.dm:78]: `spawn()` :angry:

Comment (dearmochi) [reply code/datums/progressbar.dm:78]: Since this is in `Destroy()`, I wasn't sure if a timer was going to work so I opted for that instead

Comment (SteelSlayer) [reply code/datums/progressbar.dm:78]: Timer *should* work in destroy

**Pull request closed without merging**